### PR TITLE
Scripts: Update webpack and related dependencies to the latest version

### LINKED ISSUE #53552: @webpack-cli/generators@3.0.7 has some deprecated NPM modules
### Description

When `ncu` was executed, `@wordpress/scripts` was displayed as `^26.9.0 → ^26.10.0`, so `ncu -u` `npm install` was executed in sequence.
As a result, several npm modules were shown as deprecated as follows:

npm WARN deprecated debuglog@1.0.1: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.

npm WARN deprecated readdir-scoped-modules@1.1.0: This functionality has been moved to @npmcli/fs
npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs

These issues has been resolved in [3891](https://github.com/webpack/webpack-cli/issues/3891), [3892](https://github.com/webpack/webpack-cli/issues/3892), and [3893](https://github.com/webpack/webpack-cli/issues/3893) reported in webpack-cli, so it is appropriate to replace it with the latest version.

### Step-by-step reproduction instructions

1. ncu
2. ncu -u
3. npm install

### Screenshots, screen recording, code snippet

image

image

image

## System:
 - OS: macOS 13.5
 - CPU: (12) x64 Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz
 - Memory: 478.66 MB / 32.00 GB
 - Shell: 3.2.57 - /bin/bash
## Binaries:
 - Node: 20.5.0 - ~/.nodebrew/current/bin/node
 - Yarn: 1.22.19 - /usr/local/bin/yarn
 - npm: 9.8.0 - ~/.nodebrew/current/bin/npm
## npmPackages:
 - @webpack-cli/generators: ^3.0.7 => 3.0.7 

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by gziolo**
## What?

Let's bring webpack and related dependencies to the latest version and apply them to WordPress core around 6.4 Beta 1:

- `webpack` updated to the latest version
- `browserslist` and `caniuse-lite` updated to the latest version
- `react-refresh` updated to the latest version

In addition, it tries to fix https://github.com/WordPress/gutenberg/issues/53552 caused by the outdated version of `webpack-cli`:

> When ncu was executed, @wordpress/scripts was displayed as ^26.9.0 → ^26.10.0, so ncu -u npm install was executed in sequence.
> As a result, several npm modules were shown as deprecated as follows:
> 
> npm WARN deprecated debuglog@1.0.1: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.
> npm WARN deprecated readdir-scoped-modules@1.1.0: This functionality has been moved to @npmcli/fs
> npm WARN deprecated @npmcli/move-file@1.1.2: This functionality has been moved to @npmcli/fs
> npm WARN deprecated @npmcli/move-file@2.0.1: This functionality has been moved to @npmcli/fs

## Why?

> These issues has been resolved in https://github.com/webpack/webpack-cli/issues/3891, https://github.com/webpack/webpack-cli/issues/3892, and https://github.com/webpack/webpack-cli/issues/3893 reported in webpack-cli, so it is appropriate to replace it with the latest version.

## How?

It brings `webpack-cli` to the latest version. There were some breaking changes in [v5.0.0](https://github.com/webpack/webpack-cli/releases/tag/webpack-cli%405.0.0), but they don't seem to have impact like:
-  remove the `--hot` option in favor of directly using the `HotModuleReplacement` plugin (only for build command, for serve it will work). I don't see the listed deprecated packages in the lock file anymore.

There was a recent version bump enforced for the `webpack` in the repository with https://github.com/WordPress/gutenberg/pull/49075, so here we follow-up with another version bump that also gets applied to `@wordpress/scripts`.

## Testing Instructions

1. Run `nvm use` to switch to the currently supported Node.js and npm versions.
2. Run `npm install` to install the latest version of dependencies.
3. Run `npm run build` and confirm the the build finishes correctly.
4. Run `npx wp-create-block example-static --no-wp-scripts` to confirm that Create Block still works.
5. Run `cd example-statice` to enter the newly created project.
6. Run `../node_modules/.bin/wp-scripts start --hot` to confirm that you can develop blocks with the dev server.
7. Open the site at https://localhost:8888/wp-admin, go to the post editor and confirm that you can insert Example Static block.
8. Edit the JS code of the block in the editor. It must be something controlled by React, for example `edit` function. See that the block gets automatically updated in the post editor after saving changes locally.

## Screenshots or screencast 

I confirmed that React Fast Refresh still works with the updated `webpack-cli` and other dependencies:

Screenshot 2023-09-20 at 11 56 08

### CODE REVIEW

**Comment by gziolo** [comment]
@desrosj, applying similar changes in the WordPress core would be great shortly before or even after Beta 1.

**Comment by desrosj** [comment]
I've created WordPress/wordpress-develop#5264 to accompany this change. Like this PR, there seems to be very minimal changes required, and there are no drastic changes to any built output.

**Comment by desrosj** [comment]
For the `caniuse-lite` update in Core, I'd like to split that out. I've opened WordPress/wordpress-develop#5265 to handle that.

File: packages/readable-js-assets-webpack-plugin/test/__snapshots__/build.js.snap [hunk 1]
```diff
@@ -16,7 +16,7 @@ notMinified();
 exports[`ReadableJsAssetsWebpackPlugin should produce the expected output: Asset file index.min.js should match snapshot 1`] = `"console.log("hello");"`;
 
 exports[`ReadableJsAssetsWebpackPlugin should produce the expected output: Asset file view.js should match snapshot 1`] = `
-"/******/ (() => { // webpackBootstrap
+"/******/ (function() { // webpackBootstrap
```
Comment (youknowriad) [inline packages/readable-js-assets-webpack-plugin/test/__snapshots__/build.js.snap:19]: Weird, I though arrow support was good enough :P

**Review by youknowriad — APPROVED** [review]
It's cool that we're now able to update webpack without too much hassle :)

File: packages/readable-js-assets-webpack-plugin/test/__snapshots__/build.js.snap (hunk 1 above)
Comment (gziolo) [reply packages/readable-js-assets-webpack-plugin/test/__snapshots__/build.js.snap:19]: There might be a reason why webpack wraps code with a regular function now 🤷🏻

**Pull request merged**